server/routes/activity-list/handlers/AbsencesRouteHandler.ts

```diff
@@ -0,0 +1,149 @@
+import { Request, Response } from 'express'
+import { mapToTableRow } from './activityListHelper'
+import { ActivityByLocation, AttendanceForm, CodeNameStringPair, OffenderActivityId } from '../../../@types/dps'
+import PrisonService from '../../../services/prisonService'
+import { AttendanceDto } from '../../../@types/whereaboutsApiImport/types'
+import { CreateAttendanceDtoLenient, UpdateAttendanceDtoLenient } from '../../../@types/whereaboutsApiImportCustom'
+
+function getOrInitAttendanceForm(attendanceForms: Map<string, AttendanceForm>, key: string) {
+  let attForm = attendanceForms.get(key)
+  if (!attForm) {
+    attForm = {}
+    attendanceForms.set(key, attForm)
+  }
+  return attForm
+}
+
+export default class AbsencesRouteHandler {
+  constructor(private readonly prisonService: PrisonService) {}
+
+  GET = async (req: Request, res: Response): Promise<void> => {
+    const { activityList, absenceReasons } = res.locals
+    const { activitiesNotAttended } = req.session.data
+
+    const activityAbsences = activityList.filter((a: ActivityByLocation) => {
+      return !!activitiesNotAttended.find(
+        (ana: OffenderActivityId) => ana.bookingId === a.bookingId && ana.activityId === a.eventId,
+      )
+    })
+
+    const viewContext = {
+      locationId: req.query.locationId,
+      date: req.query.date as string,
+      period: req.query.period as string,
+      rowData: activityAbsences.map(mapToTableRow),
+      paidAbsenceReasons: absenceReasons?.paidReasons,
+      unpaidAbsenceReasons: absenceReasons?.unpaidReasons.filter(
+        (r: CodeNameStringPair) => !r.code.includes('Warning'),
+      ),
+      paidAbsenceSubReasons: absenceReasons?.paidSubReasons,
+      unpaidAbsenceSubReasons: absenceReasons?.unpaidSubReasons,
+    }
+    res.render('pages/activityList/absences', viewContext)
+  }
+
+  POST = async (req: Request, res: Response): Promise<void> => {
+    const { user, activityList } = res.locals
+    const { locationId, date, period } = req.body
+    const attendanceForms = new Map<string, AttendanceForm>()
+
+    Object.entries(req.body).forEach(([key, value]) => {
```

Comment:
This block might benefit from some comments? I realise (from seeing the demo) what its doing but its not obvious from the code with the split[1-3]

Reply:
Agreed - is it ok if I make a note of this and do it on the next PR?

Reply:
Pushed this change to SAA-124